include/wil/com.h

```diff
@@ -3002,6 +3002,135 @@ namespace wil
 
 #endif // __IObjectWithSite_INTERFACE_DEFINED__
 
+
+// if C++17 or greater
+#if WIL_HAS_CXX_17
+#ifdef WIL_ENABLE_EXCEPTIONS
+namespace details
+{
+    template<typename>
+    struct com_enumerator_next_traits;
+
+    template<typename Itf, typename T, typename Ret>
+    struct com_enumerator_next_traits<Ret(__stdcall Itf::*)(ULONG, T*, ULONG*)>
+    {
+        using Interface = Itf;
+        using Result = T;
+    };
+
+
+    template<typename Interface>
+    struct com_enumerator_traits
+    {
+        using Result = typename com_enumerator_next_traits<decltype(&Interface::Next)>::Result;
+        using smart_result = std::conditional_t<
+            std::is_pointer_v<Result> &&
+            std::is_base_of_v<::IUnknown, std::remove_pointer_t<Result>>, wil::com_ptr<std::remove_pointer_t<Result>>, Result>;
+    };
+}
+
+template<typename T>
+struct is_com_ptr : std::false_type {};
+
+template<typename T>
+struct is_com_ptr<wil::com_ptr<T>> : std::true_type {};
+
+template <typename IEnumType, typename TStoredType = typename details::com_enumerator_traits<IEnumType>::smart_result>
+struct iterator
+{
+    wil::com_ptr<IEnumType> m_enum{};
+    TStoredType m_currentValue{};
+
+    static constexpr bool is_com_ptr_v = is_com_ptr<TStoredType>::value;
+
+    iterator(iterator&&) = default;
+    iterator(iterator const&) = default;
+    iterator& operator=(iterator&&) = default;
+    iterator& operator=(iterator const&) = default;
+
+    iterator(IEnumType* enumPtr) : m_enum(enumPtr)
+    {
+        FetchNext();
+    }
+
+    auto operator->()
+    {
+        return wistd::addressof(m_currentValue);
+    }
+
+    auto& operator*()
+    {
+        return m_currentValue;
+    }
+
+    iterator& operator++()
+    {
+        // If we're already at the end, don't try to advance. Otherwise, use Next to advance.
+        if (m_enum)
+        {
+            FetchNext();
+        }
+
+        return *this;
+    }
+
+    bool operator!=(iterator const& other) const
+    {
+        return !(*this == other);
+    }
+
+    bool operator==(iterator const& other) const
+    {
+        return (m_enum.get() == other.m_enum.get());
+    }
+
+private:
+    void FetchNext()
+    {
+        if (m_enum)
+        {
+            m_currentValue = TStoredType{};
+            auto hr = m_enum->Next(1, &m_currentValue, nullptr);
+            if (hr == S_FALSE)
+            {
+                m_enum = nullptr;
+            }
+            else 
+            {
+                THROW_IF_FAILED_MSG(hr, "Failed to get next");
+            }
+        }
+    }
+};
+
+template<typename IEnumXXX>
+struct iterator_range

```

Comment:
Should this type be declared inside make_range so it's not just hanging around visible here?

Similarly with the iterator<T> - if it's local to make_range then you have to obtain one from make_range...

Reply:
I'll move the `iterator_range` type inside `make_range`. I'd like to keep the iterator itself visible - esp. as it allows the types in the iterator to be tested, lmk if you disagree.